### PR TITLE
Add an option (-r) to specify alternate root directory

### PR DESCRIPTION
**Pull request by jose1711**
This may come handy when you have a different system available inside
a separate mountpoint.

### CODE REVIEW

File: fakepkg [hunk 1]
```diff
@@ -55,19 +56,15 @@ function fakebuild() {
 
 	# Assembling a list of all files belonging to the package
 	# The tar command is unable to handle very many arguments therefore create a file
-	pacman -Qlq $1 > ./package_file_list
-	echo ".PKGINFO" >> ./package_file_list
-	[[ -f ".INSTALL" ]] && echo ".INSTALL" >> ./package_file_list
-	[[ -f ".Changelog" ]] && echo ".Changelog" >> ./package_file_list
-	[[ -f ".MTREE" ]] && echo ".MTREE" >> ./package_file_list
+	pacman --dbpath "${ROOT_DIR}/var/lib/pacman" -Qlq $1 | sed 's%^/%%' > ./package_file_list
```
Comment (Edenhofer) [inline fakepkg]: Why not use the `--sysroot` option here instead as it would save us from having to hard-code the DB path?
At least make it some variable and change line 15 as well as to make the path somewhat configurable.

File: fakepkg [hunk 2]
```diff
@@ -10,9 +10,10 @@ declare -r tmp_root='fakepkg'
 # Actual assembly function
 # $1: pkgname , $2: destination dir , $3 verbosity
 function fakebuild() {
+	pkgver=$(pacman -Qi --dbpath "${ROOT_DIR}/var/lib/pacman" ${package} 2>/dev/null | sed -n 2p | awk '{print $3}')
```
Comment (Edenhofer) [inline fakepkg]: Sorry, this is completely unrelated and originated from my initial first attempt, but the awk plus sed stuff can be written more neatly as `sed -n -e 's/^Version\s*: \(.*\)/\1/p'`.

File: fakepkg [hunk 3]
```diff
@@ -142,11 +141,13 @@ for ARG in "$@"; do
 done
 
 # Assembling PKG_LIST and DEST_DIR
-PARAMS=$(getopt -o o:j:vV --long out:,jobs:,verbose,version -n "$myname" -- "$@")
+PARAMS=$(getopt -o o:j:r:v --long out:,jobs:,root:,verbose -n "$myname" -- "$@")
```
Comment (Edenhofer) [inline fakepkg]: Is there a reason for leaving out the `Version` option or did this just happen by accident?

File: fakepkg [hunk 4]
```diff
@@ -165,6 +166,13 @@ while true ; do
 		shift
 		;;
 
+		-r|--root)
+		ROOT_DIR=$2
+		[[ ! -d "${ROOT_DIR}" ]] && echo -e "The specified alternate root directory does not exist!" && exit -3
+		ROOT_DIR=$(cd "${ROOT_DIR}"; pwd)
```
Comment (Edenhofer) [inline fakepkg]: `readlink -e "${ROOT_DIR}"` should work as well and would get rid of a subshell invocation

File: fakepkg [hunk 5]
```diff
@@ -55,19 +56,15 @@ function fakebuild() {
...
-	[[ -f ".INSTALL" ]] && echo ".INSTALL" >> ./package_file_list
-	[[ -f ".Changelog" ]] && echo ".Changelog" >> ./package_file_list
-	[[ -f ".MTREE" ]] && echo ".MTREE" >> ./package_file_list
+	pacman --dbpath "${ROOT_DIR}/var/lib/pacman" -Qlq $1 | sed 's%^/%%' > ./package_file_list
 
 	# Taring things together
 	if [[ $3 -ge "1" ]]; then
 		#XZ_OPT="-T 0" \
-		tar --ignore-failed-read --owner=0 --group=0 --no-recursion -c -a --force-local -f "${PKG_NAME}.pkg.tar.${PKG_EXT}" -T ./package_file_list
+		tar -C "${ROOT_DIR:-/}" --ignore-failed-read --owner=0 --group=0 --no-recursion -c -a --force-local -f "${PKG_NAME}.pkg.tar.${PKG_EXT}" -T ./package_file_list -C "$TMPDIR" .PKGINFO .MTREE .INSTALL .Changelog
```
Comment (Edenhofer) [inline fakepkg]: I doubt that specifying `-C` twice works as intended

File: fakepkg [hunk 6]
```diff
@@ -55,19 +56,15 @@ function fakebuild() {
 
 	# Assembling a list of all files belonging to the package
 	# The tar command is unable to handle very many arguments therefore create a file
-	pacman -Qlq $1 > ./package_file_list
-	echo ".PKGINFO" >> ./package_file_list
-	[[ -f ".INSTALL" ]] && echo ".INSTALL" >> ./package_file_list
-	[[ -f ".Changelog" ]] && echo ".Changelog" >> ./package_file_list
```
Comment (Edenhofer) [inline fakepkg:61]: This change is pretty much independent of the rest of the changes and hence should be in its own commit.
However, I think it does not work as intended as `tar` throws errors if it receives files as input which do not exist. Note, `.INSTALL`, ... must not exist and are only added to the list of files which to back up if they really are present.

File: fakepkg [hunk 7]
```diff
@@ -191,7 +199,8 @@ PKG_LIST=($*)
 # Checking the PKG_LIST
 [[ ${#PKG_LIST[@]} == 0 ]] && usage && exit 1
 for package in "${PKG_LIST[@]}"; do
-	if [[ ! -d /var/lib/pacman/local/${package}-$(pacman -Qi ${package} 2>/dev/null | sed -n 2p | awk '{print $3}') ]]; then
+	pkgver=$(pacman -Qi --dbpath "${ROOT_DIR}/var/lib/pacman" ${package} 2>/dev/null | sed -n 2p | awk '{print $3}')
```
Comment (Edenhofer) [inline fakepkg]: see above

**Review by Edenhofer — CHANGES_REQUESTED** [review]

File: fakepkg [hunk 8]
```diff
@@ -108,10 +105,12 @@ function usage() {
 	    -v, --verbose           Increase verbosity
 	    -j, --jobs <jobs>       Build in parallel - you may want to set XZ_OPT
 	    -o, --out  <dir>        Write output to <dir>
+	    -r, --root <dir>        Use an alternate root directory containing pacman db
```
Comment (Edenhofer) [inline fakepkg:116]: Please also amend the README and the man page

File: fakepkg (hunk 1 above)
Comment (jose1711) [reply fakepkg]: `--sysroot` does look compelling. OTOH it does require root privileges which `--dbpath` approach might not need.

File: fakepkg (hunk 2 above)
Comment (jose1711) [reply fakepkg]: Actually, I was thinking more at `awk 'NR == 2{print $3}'`. Not a big fan of escapes.

File: fakepkg (hunk 3 above)
Comment (jose1711) [reply fakepkg]: Sorry, yes, my bad. Actually I was doing the original changes to the older version as you did not bump the version since the last modifications. Then realizing the issue I was redoing the changes against the master. And clearly missed this one..

File: fakepkg (hunk 5 above)
Comment (jose1711) [reply fakepkg]: and yet..
```
$ /mkdir /tmp/{a,b}
$ touch /tmp/a/1 /tmp/b/2
$ tar -cvf /tmp/foo.tar -C /tmp/a 1 -C /tmp/b 2 
1
2
$ tar tvf /tmp/foo.tar 
-rw-r--r-- jose/users        0 2019-01-02 00:57 1
-rw-r--r-- jose/users        0 2019-01-02 00:57 2
```

File: fakepkg (hunk 6 above)
Comment (jose1711) [reply fakepkg:61]: Yes, I did notice this one. However getting only warning (and only in verbose mode) I did not considered it to be the real issue. I will put the metadata related files into an array so `tar` won't be bothered by nonexisting files.

File: fakepkg (hunk 8 above)
Comment (jose1711) [reply fakepkg:116]: (Another) good point.

File: fakepkg (hunk 6 above)
Comment (Edenhofer) [reply fakepkg:61]: An array is fine, however if there already is a file which contains all the other paths, why not simply amend it like it is done now?

File: fakepkg (hunk 5 above)
Comment (Edenhofer) [reply fakepkg]: Thanks a bunch for the correction! I previously just took a look at the man page and though it would not be possible.

Either way, like I mentioned below, this change is somewhat independent from introducing a `-r` option and belongs into its own commit. However, to be completely honest, I would prefer to just amend the paths to `package_file_list`.

File: fakepkg (hunk 3 above)
Comment (Edenhofer) [reply fakepkg]: Ah, I see. No problem.

File: fakepkg (hunk 2 above)
Comment (Edenhofer) [reply fakepkg]: Admittedly, the escapes are ugly. However, searching for `Version` instead of hard-coding line 2, is prettier in my opinion.

File: fakepkg (hunk 1 above)
Comment (Edenhofer) [reply fakepkg]: You are right, requiring root would not be ideal. Just stick to specifying `--dpath`.

Regardless of the used option, may I ask you to put `/var/lib/pacman/` into a variable and use it throughout the script.

File: fakepkg (hunk 2 above)
Comment (jose1711) [reply fakepkg]: Without a doubt. OTOH.. do we want to set `LC_MESSAGES` to `C.utf8` for this to work universally? Some users may use non-English locales even for `root` (not me :smile:)

File: fakepkg (hunk 5 above)
Comment (jose1711) [reply fakepkg]: This might've seen as a separate change but in fact it is not. With a default `ROOT_PATH` set to `/` `tar` has no troubles finding the files as `pacman -Qlq` output tells `tar` exactly where it should look for source files. This changes when you have to start elsewhere - `pacman` will give you `/usr/share/blabla` but you need to first change into your new `ROOT_PATH` - hence the `-C`. At the same time you need to keep `.PKGINFO`, `.MTREE` etc at the "top" of the archive.

File: fakepkg (hunk 6 above)
Comment (jose1711) [reply fakepkg:61]: I could not find any other way how to put both `$ROOT_PATH/somedir/somefile` and `.PKGINFO` at the top of the archive while at the same time not showing the `$ROOT_PATH` component in the archive.

File: fakepkg (hunk 2 above)
Comment (Edenhofer) [reply fakepkg]: Very good point :see_no_evil: indeed there might be such users...
I guess `LANG=C` should be sufficient though.

File: fakepkg (hunk 5 above)
Comment (Edenhofer) [reply fakepkg]: Right, agreed.

File: fakepkg (hunk 6 above)
Comment (Edenhofer) [reply fakepkg:61]: Sorry for the noise, I admittedly did not think of that.

**Review by Edenhofer — APPROVED** [review]
Thanks a lot for putting in the work to implement this feature. Thanks also for the surrounding discussion and for bearing with me.

**Comment by Edenhofer** [comment]
Thanks again!
I will tag a new release in the coming days.

**Comment by jose1711** [comment]
> Thanks again!
> I will tag a new release in the coming days.

my pleasure. it was a very nice experience. thank you and all the best in 2019!

**Comment by Edenhofer** [comment]
I am glad to hear that! You too!
Btw. I just release version 1.41.0, see [here](https://github.com/Edenhofer/fakepkg/releases/tag/v1.41.0).

However, you forgot to properly test your changes. Some clean ups were necessary after the merge as you misused the variable "$package", see commit b5080ef00b8872dc6d190a36da3c2f3540506eda. Please be sure to properly test your changes in the future.
